source/extensions/transport_sockets/tap/tap_config_impl.cc

```diff
@@ -106,31 +107,68 @@ bool PerSocketTapperImpl::shouldSendStreamedMsgByConfiguredSize() const {
   return config_->minStreamedSentBytes() > 0;
 }
 
-void PerSocketTapperImpl::handleSendingStreamTappedMsgPerConfigSize(const Buffer::Instance& data,
-                                                                    const uint32_t total_bytes,
-                                                                    const bool is_read,
-                                                                    const bool is_end_stream) {
+void PerSocketTapperImpl::submitStreamedDataPerConfiguredSize() {
+  sink_handle_->submitTrace(std::move(streamed_trace_));
+  streamed_trace_.reset();
+  current_streamed_rx_tx_bytes_ = 0;
+}
+
+void PerSocketTapperImpl::setStreamedDataPerConfiguredSize(const Buffer::Instance& data,
+                                                           const uint32_t buffer_offset,
+                                                           const uint32_t total_bytes,
+                                                           const bool is_read,
+                                                           const bool is_end_stream) {
+
   makeStreamedTraceIfNeeded();
   auto& event =
       *streamed_trace_->mutable_socket_streamed_trace_segment()->mutable_events()->add_events();
   initStreamingEvent(event);
-  uint32_t buffer_start_offset = 0;
   if (is_read) {
-    buffer_start_offset = data.length() - total_bytes;
     TapCommon::Utility::addBufferToProtoBytes(*event.mutable_read()->mutable_data(), total_bytes,
-                                              data, buffer_start_offset, total_bytes);
+                                              data, buffer_offset, total_bytes);
     current_streamed_rx_tx_bytes_ += event.read().data().as_bytes().size();
   } else {
-    event.mutable_write()->set_end_stream(is_end_stream);
     TapCommon::Utility::addBufferToProtoBytes(*event.mutable_write()->mutable_data(), total_bytes,
-                                              data, buffer_start_offset, total_bytes);
+                                              data, buffer_offset, total_bytes);
+    event.mutable_write()->set_end_stream(is_end_stream);
     current_streamed_rx_tx_bytes_ += event.write().data().as_bytes().size();
   }
+}
```

Comment:
Sorry I may didn't make it clear. I mean all these changes are unnecessary. You can revert almost all changes and only keep one method to do the duration check and add the condition at the `handleSendingStreamTappedMsgPerConfigSize`.